predicators/datasets/generate_atom_trajs_with_vlm.py

```diff
@@ -173,33 +174,64 @@ def _sample_vlm_atom_proposals_from_trajectories(
+                                                         num_completions=1)
+        assert len(curr_vlm_atom_labelling) == 1
+        sanitized_output = curr_vlm_atom_labelling[0].replace('\\', '')
+        curr_traj_txt_outputs.append(sanitized_output)
+        curr_scenes_labelled += 1
+        logging.info(
+            f"Completed ({curr_scenes_labelled}/{total_scenes_to_label}) " \
+            f"label queries to VLM for trajectory #{idx + 1} of {num_trajs}!"
+        )
+    logging.info("Finished labeling trajectory.")
+    return curr_traj_txt_outputs
+
+
 def _label_trajectories_with_vlm_atom_values(
         trajectories: List[ImageOptionTrajectory], vlm: VisionLanguageModel,
         atoms_list: List[str]) -> List[List[str]]:
     """Given a list of atoms, label every state in ImageOptionTrajectories with
-    the truth values of a set of atoms."""
-    total_scenes_to_label = sum(len(traj.imgs) for traj in trajectories)
-    curr_scenes_labelled = 0
+    the truth values of those atoms."""
     output_labelled_atoms_txt_list = []
-    for traj in trajectories:
-        curr_traj_txt_outputs: List[str] = []
-        prompts_for_traj = _generate_prompt_for_scene_labelling(
-            traj, atoms_list, label_history=curr_traj_txt_outputs)
-        for text_prompt, img_prompt in prompts_for_traj:
-            # Sample VLM outputs with temperature 0 in an attempt to be
-            # accurate.
-            curr_vlm_atom_labelling = vlm.sample_completions(text_prompt,
-                                                             img_prompt,
-                                                             0.0,
-                                                             CFG.seed,
-                                                             num_completions=1)
-            assert len(curr_vlm_atom_labelling) == 1
-            sanitized_output = curr_vlm_atom_labelling[0].replace('\\', '')
-            curr_traj_txt_outputs.append(sanitized_output)
-            curr_scenes_labelled += 1
-            logging.info("Completed (%s/%s) label queries to VLM!",
-                         curr_scenes_labelled, total_scenes_to_label)
-        output_labelled_atoms_txt_list.append(curr_traj_txt_outputs)
+
+    def label_function(
+            indexed_traj: Tuple[int, ImageOptionTrajectory]) -> List[str]:
+        return _label_single_trajectory_with_vlm_atom_values(
+            indexed_traj, vlm, atoms_list, len(trajectories))
+
+    indexed_trajectories = list(enumerate(trajectories))
+
+    if CFG.grammar_search_parallelize_vlm_labeling:
+        logging.info("Labeling trajectories in parallel.")
+        with ThreadPoolExecutor() as executor:
+            for traj_txt_outputs in executor.map(label_function,
+                                                 indexed_trajectories):
+                output_labelled_atoms_txt_list.append(traj_txt_outputs)
+        return output_labelled_atoms_txt_list
```

Comment:
very cool! Is there any way to control the number of parallel processes we fork? Or will the system just fork as many as it can? Would be nice to introduce a command line arg that dictates this if possible, otherwise a comment explaining what's happening with the number of forks would be good!

Reply:
This is only multithreading, not multiprocessing. Since the thing we need to parallelize here is a network I/O thing basically (waiting for response back from google's servers), multiprocessing isn't particularly helpful -- multithreading is sufficient. I think we make 1 thread for each trajectory, and we can't do better than that.